File: CS456/A3/Router/router.rs
mod helpers;
mod packet;
mod state;

use std::{env, net, thread};
use std::fs::File;
use std::io::Write;
use packet::Packet;

fn main() {
    /*
    Argument validation and parsing
    */

    let args: Vec<_> = env::args().collect();

    if args.len() != 5 {
        helpers::abort("Expects arguments: <router_id>, <nse_hostname>, <router_port>, <nse_port>");
    }

    let router_id: u8 = args[1].parse::<u8>().expect("Invalid router_id");
    let s_router_id: &str = &*args[1].to_string();
    let s_router_name: String = "R".to_string() + s_router_id;

    let nse_hostname: &str = &*args[2].to_string();
    let router_port: u16 = args[3].parse::<u16>().expect("Invalid router_port");
    let nse_port: u16 = args[4].parse::<u16>().expect("Invalid nse_port");

    /*
    Initialization
    */
    let socket = net::UdpSocket::bind(("localhost", router_port)).expect("Error binding to socket");

    let log_file_name = "router".to_string() + s_router_id + ".log";
    let mut log = File::create(&log_file_name).expect("Unable to create log file");

    /*
    Main Program
    */
    macro_rules! udp_send {
        ($packet: expr) => {
            let pkt = $packet;
            let pkt_bits = pkt.encode();
            let log_msg = s_router_name + " " + &*pkt.log_message();

            socket.send_to(&pkt_bits, (nse_hostname, nse_port)).expect("Failed to send packet");

            log.write_all(&*log_msg.as_bytes()).expect("Failed to write log");
        }
    }

    // 1) Send Init

    udp_send!(packet::Init{router_id: router_id});


    // 2) Send Hello to neighbours


    // 3) Receive Hello and LSPDU and flood LSPDU


}

File: CS456/A3/Router/router.rb
require 'socket'

#
# Parsing arguments
#
raise RuntimeError, 'Expects arguments: <router_id>, <nse_hostname>, <nse_port>, <router_port>' unless ARGV.length == 4

ROUTER_ID = ARGV[0].to_i
NSE_HOSTNAME = ARGV[1]
NSE_PORT = ARGV[2].to_i
ROUTER_PORT = ARGV[3].to_i

raise RuntimeError, 'Invalid router_id' if ROUTER_ID == 0
raise RuntimeError, 'Invalid nse_port' if NSE_PORT == 0 or NSE_PORT > 0xffff
raise RuntimeError, 'Invalid router_port' if ROUTER_PORT == 0 or ROUTER_PORT > 0xffff

#
# Initialization
#
socket = UDPSocket.new
socket.bind '', ROUTER_PORT

log_file = open("router#{ROUTER_ID}.log", 'w')

#
# Main program
#
load 'packet.rb'
load 'circuit.rb'
load 'graph.rb'

# 1 - Send INIT
PacketInit.new(ROUTER_ID).udp_send(socket,log_file)


# 2 - Receive circuit DB from nse
rcd = socket.recv(64).unpack('L*')
circuit_db = CircuitDB.new

(0..rcd[0]-1).each do |i|
  circuit_db[rcd[1+(i*2)]] = rcd[2+(i*2)]
end

log_file.write("\nR#{ROUTER_ID} receives circuit_db #{circuit_db}\n\n")


# 3 - Send HELLO to all neighbours
circuit_db.keys.each do |link|
  PacketHello.new(ROUTER_ID, link).udp_send(socket, log_file)
end
log_file.write("\n")

# 4 - Listen for HELLO and LSPDU
loop do
  packet = make_pkt(socket.recv(64).unpack('L*'))
  log_file.write("R#{ROUTER_ID} receives #{packet}\n")
  packet.deliver
end
